src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,119 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size = 0;
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value, tail, null);
+        if (size == 0) {
+            head = newNode;
+        } else {
+            tail.setNext(newNode);
+        }
+        tail = newNode;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            Node<T> newNode = new Node<>(value, null, head);
+            head.setPrev(newNode);
+            head = newNode;
+        } else {
+            Node<T> current = getNodeAt(index);
+            Node<T> newNode = new Node<>(value, current.getPrev(), current);
+            if (current.getPrev() != null) {
+                current.getPrev().setNext(newNode);
+            }
+            current.setPrev(newNode);
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNodeAt(index).getItem();
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> current = getNodeAt(index);
+        T oldValue = current.getItem();
+        current.setItem(value);
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToRemove = getNodeAt(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.getItem();
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if ((object == null && current.getItem() == null)
+                    || (object != null && object.equals(current.getItem()))) {
+                unlink(current);
+                return true;
+            }
+            current = current.getNext();
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeAt(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.getNext();
+        }
+        return current;
```

Comment:
The `getNodeAt` method should be optimized to iterate only half of the list based on the index. If the index is in the first half, start from the head; if it's in the second half, start from the tail. This will significantly improve performance for large lists.